### PR TITLE
MGMT-4872: Initial events generation

### PR DESCRIPTION
**Pull request by masayag**
Events generation is designed to expose a uniformed method of
initializing the assisted-service events, based on each event definition.
The event definition will require a set of parameters for each event type.
It will be used for generating a function for initializing the event with
the provided parameters, hence enforcing to provide the parameters at
initialization and not to miss any needed information.
The events will be defined in `./data/events.yaml`.

In order to introduce a new event, the developer will have to follow the
next steps:
1. Add event definition to `data/events.yaml` (event definition will be
   discussed ahead).
2. Generate the code for creating the event by: `skipper make generate
   generate-events`
3. Use the generated function for the event from
   `./internal/common/events.event.go` for creating the event to be
   emitted by the event handler.

Event definition should specify the following attributes:

-  name: The name of the event which needs to remain unique and constant
   as it may be referred by the service's clients (e.g. by the UI).
- message: A template of the message that will be rendered if it
  contains references to the properties. E.g. the message "Install
  cluster {cluster_id}" expects the existence of a property named
  cluster_id
- event_type: Can be either "cluster" or "host". "cluster" type requires
  the existence of "cluster_id" in properties and "host" type requires
  the existence of "host_id".
- severity: Any of info, warning, error or critical
- properties: A list of properties to be rendered into the message (if
  referred by) or added as identifier of the message (e.g. cluster_id, host_id)
  with the type of the property.

Event's severity is defined as a field of the event definition in ./data/events.yaml.
In special events (e.g. generic events) where the severity value of the event is set
by the caller, the severity value can be provided as a property.
In this case a priority will be given to the severity value provided as a property.

### CODE REVIEW

File: data/events.yaml
Comment (ronniel1) [inline data/events.yaml:2]: @avishayt do we need event IDs or is the name enough?

Comment (masayag) [reply data/events.yaml:2]: I removed the ID since I haven't noticed it is being used or persisted to the events table.
Unless it is required for a specific use, I'd postpone adding it to the future if/when it will be needed in order not to commit to it.

Comment (avishayt) [reply data/events.yaml:2]: Agree with @masayag.
However, we need to be conscious of the fact that users (including internal users like UI and operator) will rely on the name as an ID.  We'll need to maintain that contract.

Comment (ronniel1) [reply data/events.yaml:2]: OK

**Comment by eranco74** [comment]
@masayag is this still relevant?

**Comment by masayag** [comment]
> @masayag is this still relevant?

Yes, it is currently being worked on by @aireni.

File: data/events.yaml
Comment (masayag) [inline data/events.yaml:26]: @arielireni let's maintain a common indentation of 2 spaces offset

File: data/events.yaml
Comment (masayag) [inline data/events.yaml:67]: let's unify `log_message` and `message` into `message`
if both exists, prefer the value of  `log_message` for the `message` attribute

File: data/events.yaml
Comment (masayag) [inline data/events.yaml:504]: let's add documentation in commit message (the first commit) for the severity field which is allowed to override the event.severity field.

File: internal/bminventory/inventory.go
Comment (masayag) [inline internal/bminventory/inventory.go:2772]: no need to include the message itself, only provide the `hostutil.GetHostnameForMsg(host)` as value for the host_name

File: internal/bminventory/inventory.go
Comment (masayag) [inline internal/bminventory/inventory.go:3250]: pls move one line above to preserve the original order (log before event)

File: internal/bminventory/inventory.go
Comment (masayag) [inline internal/bminventory/inventory.go:4861]: pls remove, it repeats of the previous comment line

File: internal/bminventory/inventory.go
Comment (masayag) [inline internal/bminventory/inventory.go:4882]: pls remove, it repeats of the previous comment line

File: internal/bminventory/inventory_test.go
Comment (masayag) [inline internal/bminventory/inventory_test.go:7225]: pls delete commented block

File: internal/events/event.go
Comment (masayag) [inline internal/events/event.go:28]: It seems that > 95% of the time we call this method with time.Now().
What do you think about adding a method that doesn't expect eventTime and fires it immediately ? 
e.g 
```
SendClusterEvent(ctx context.Context, event ClusterEvent)
SendClusterEventAtTime(ctx context.Context, event ClusterEvent, eventTime time.Time)
```
same for host-events

File: internal/host/transition_test.go
Comment (masayag) [inline internal/host/transition_test.go:529]: pls delete commented code

File: internal/host/transition_test.go
Comment (masayag) [inline internal/host/transition_test.go:4056]: line numbers might be changed in the future (as code evolves), pls rephrase the comment.

File: internal/host/transition_test.go
Comment (masayag) [inline internal/host/transition_test.go:4818]: pls remove commented code

File: tools/generate_events.py
Comment (masayag) [inline tools/generate_events.py:2]: pls delete spare empty line

File: tools/generate_events.py
Comment (masayag) [inline tools/generate_events.py:128]: pls delete commented lines

File: tools/generate_events.py
Comment (masayag) [inline tools/generate_events.py:129]: pls delete 2 commented lines

File: data/events.yaml
Comment (arielireni) [reply data/events.yaml:504]: solved

File: internal/bminventory/inventory.go
Comment (arielireni) [reply internal/bminventory/inventory.go:2772]: solved

File: internal/bminventory/inventory.go
Comment (arielireni) [reply internal/bminventory/inventory.go:3250]: solved.
does logs before actions is a convention?

File: internal/bminventory/inventory.go
Comment (arielireni) [reply internal/bminventory/inventory.go:4861]: done

File: internal/bminventory/inventory.go
Comment (masayag) [reply internal/bminventory/inventory.go:3250]: no, there are other places in which the order is reversed.
but in this PR we'd like to preserve same behavior as used to be before.
if someone would decide that such a change is required, that can be added in a separate PR which isn't a concern of the the current one.

File: internal/bminventory/inventory.go
Comment (arielireni) [reply internal/bminventory/inventory.go:4882]: done

File: internal/bminventory/inventory_test.go
Comment (arielireni) [reply internal/bminventory/inventory_test.go:7225]: done

File: internal/host/transition_test.go
Comment (arielireni) [reply internal/host/transition_test.go:529]: done

File: internal/host/transition_test.go
Comment (arielireni) [reply internal/host/transition_test.go:4056]: solved

File: internal/host/transition_test.go
Comment (arielireni) [reply internal/host/transition_test.go:4818]: done

File: data/events.yaml
Comment (arielireni) [reply data/events.yaml:26]: solved

File: internal/events/event.go
Comment (arielireni) [reply internal/events/event.go:28]: sounds good.
in the test files it's always gomock.Any(). should I create also sendClusterEventAtAnyTime() for the tests?
(same for host)

Comment (arielireni) [reply internal/events/event.go:28]: solved.
the default time for the mock events is gomock.Any().
also, I think time.Now() is the time for the real events and gomock.Any() is the time for the mock events in 100% of the cases so I'm not sure why we need to keep the option of time as an input

File: tools/generate_events.py
Comment (arielireni) [reply tools/generate_events.py:2]: done

File: tools/generate_events.py
Comment (arielireni) [reply tools/generate_events.py:122]: done

File: tools/generate_events.py
Comment (arielireni) [reply tools/generate_events.py:123]: done

File: data/events.yaml
Comment (arielireni) [reply data/events.yaml]: solved

**Comment by masayag** [comment]
There are subsystem tests failures, some of them are justified since they are relying on event matcher and messages which seems less relevant, e.g. https://github.com/openshift/assisted-service/blob/master/subsystem/kubeapi_test.go#L2389

Let's replace the event matcher from that file with the one introduced by this PR and rely on event-type and not on the expected message

File: internal/bminventory/inventory.go
Comment (masayag) [inline internal/bminventory/inventory.go:771]: pls remove commented-out code.

File: internal/bminventory/inventory.go
Comment (masayag) [inline internal/bminventory/inventory.go:779]: same here

File: internal/bminventory/inventory.go
Comment (masayag) [inline internal/bminventory/inventory.go:790]: and here...

File: internal/bminventory/inventory.go
Comment (arielireni) [reply internal/bminventory/inventory.go:771]: done

File: internal/bminventory/inventory.go
Comment (arielireni) [reply internal/bminventory/inventory.go:779]: done

File: internal/bminventory/inventory.go
Comment (arielireni) [reply internal/bminventory/inventory.go:790]: done. sorry

File: tools/generate_events.py
Comment (eranco74) [inline tools/generate_events.py:128]: Why not just
```suggestion
    return self.event.severity.lower() 
```

**Review by eranco74 — COMMENTED** [review]
/lgtm

File: internal/events/event.go
Comment (filanov) [inline internal/events/event.go:194]: We are adding late binding feature, some host events are related to infra env and not to a specific cluster. 
cc: @yevgeny-shnaidman

File: tools/generate_events.py
Comment (arielireni) [reply tools/generate_events.py]: Fixed and added validation of supported severity values in 'validate_event' function.

File: internal/bminventory/inventory.go
Comment (carbonin) [inline internal/bminventory/inventory.go:296]: This pattern is repeated everywhere. Can we do something to make it a one line operation? Maybe add an events handler as a parameter to the generated functions and send the event from there?

Is there ever a use case to create an event and not send it?

Comment (filanov) [reply internal/bminventory/inventory.go:296]: +1

**Comment by filanov** [comment]
Maybe this effort should be split? into multiple PRs? 
For example the generation tools and few simple example of a code that will probably don't change much like enable disable host, register cluster so we can see the flow and then we can separate another PR (or multiple prs) that will switch existing events.

File: internal/bminventory/inventory.go
Comment (arielireni) [reply internal/bminventory/inventory.go]: Replaced this pattern.
Can you please re-review it? There are few examples in the updated PR.

**Comment by arielireni** [comment]
@filanov In this version of the PR, there are the core aspects (the generation tools and few simple examples of events creation). Once approved and merged, I'll send a series of PRs to replace other events with the new method. 
Can you please re-review it?

File: internal/events/events_test_utils.go
Comment (carbonin) [inline internal/events/events_test_utils.go:1]: Don't block merge on this, but do we have a convention for test utils?
I don't particularly like adding too much to the "real" package just for test utils.
Especially not exported functions and types.

Not sure who to ask about this, @filanov maybe?

File: internal/events/event.go
Comment (carbonin) [reply internal/events/event.go:194]: `GetClusterID` is really a convenience here though, right? Once we have the host we could get the cluster easily enough. Alternatively `GetClusterID` could return `nil` :man_shrugging: Maybe it needs an error return value as well?

File: tools/generate_events.py
Comment (carbonin) [inline tools/generate_events.py:98]: Regarding the late-binding conversation above ... Maybe these should be conditional on the properties provided rather than dictated explicitly based on the event type? Then things would be easier to change if/when we need a host event without a cluster id.

Comment (carbonin) [reply tools/generate_events.py:98]: Any thoughts on this one? @filanov ?

Comment (masayag) [reply tools/generate_events.py:98]: @carbonin thanks for bringing this option. 
The developer that adds the event should know the context of the event, and based on that context to define the type of the event.

If we'll have two types of host-events (in context of infraenv or cluster), we won't have to rely on existence of certain properties to determine the context of the event and which methods it should expose. 

That can be achieved by having the following types for host event:

- host - the common host event type (as within the current PR), will generate event that implements the HostEvent interface
  - BaseEvent (GetName, GetSeverity, FormatMessage)
  - GetHostID
  - GetClusterID

- lb_host - the late-binding host event type, will generate event that implements the LBHostEvent (open for other naming) interface:
  - BaseEvent (GetName, GetSeverity, FormatMessage)
  - GetHostID
  - GetInfraEnvID
  - GetClusterID // optional

Having an explicit type for LB will allow to validate the correctness of the properties specified for instantiating the event. E.g, if the type is lb_host, it is required to include the InfraEnvID, the same way the validation checks that host event type requires clusterID as a property (as in [validate_events](https://github.com/openshift/assisted-service/pull/1854/files#diff-47228a8d6bca4abbc484035491d0e5e8db24eafdd4083e1ad8373637132bfe95R213))

Since events are being added once, the cost of specifying is minor and the outcome is an expected event that enforces the caller to provide all of the required properties.

WDYT ? 
Also, do you see any reason for this discussion to block this PR ?

Comment (carbonin) [reply tools/generate_events.py:98]: > Also, do you see any reason for this discussion to block this PR ?

If the solution was to change the initial design, then yes. But I think adding a separate event type as a part of the late binding effort works. The naming can still be up for debate, but as long as we have an option that's just an additive change then this should be okay as is.

**Review by carbonin — APPROVED** [review]

File: internal/events/event.go
Comment (avishayt) [reply internal/events/event.go:194]: For late binding we would have another event type, UnboundHostEvent, right?

**Review by avishayt — APPROVED** [review]

**Pull request merged**
